stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -62,35 +81,96 @@
-    (match (map-get? rounds reward-cycle)
-        last-round (and (> round last-round) (map-set rounds reward-cycle round))
-        (map-set rounds reward-cycle round)))
+    (ok (match (map-get? rounds reward-cycle)
+        last-round (begin
+            (asserts! (<= round (+ last-round u1)) (err ERR_INVALID_ROUND))
+            (and (> round last-round) (map-set rounds reward-cycle round)))
+        (map-set rounds reward-cycle round))))
+
+;; Signer vote for the aggregate public key of the next reward cycle
+;;  Each signer votes for the aggregate public key for the next reward cycle.
+;;  This vote must happen after the list of signers has been set by the node,
+;;  which occurs in the first block of the prepare phase. The vote is concluded
+;;  when the threshold of `threshold-consensus / 1000` is reached for a
+;;  specific aggregate public key. The vote is weighted by the amount of
+;;  reward slots that the signer controls in the next reward cycle. The vote
+;;  may require multiple rounds to reach consensus, but once consensus is
+;;  reached, later rounds will be ignored.
+;;
+;;  Arguments:
+;;   * signer-index: the index of the calling signer in the signer set (from
+;;     `get-signers` in the .signers contract)
+;;   * key: the aggregate public key that this vote is in support of
+;;   * round: the voting round for which this vote is intended
+;;   * reward-cycle: the reward cycle for which this vote is intended
+;;  Returns:
+;;   * `(ok true)` if the vote was successful
+;;   * `(err <code>)` if the vote was not successful (see errors above)
+(define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint) (reward-cycle uint))
+    (let ((tally-key {reward-cycle: reward-cycle, round: round, aggregate-public-key: key})
+            ;; vote by signer weight
+            (signer-weight (try! (get-current-signer-weight signer-index)))
+            (new-total (+ signer-weight (default-to u0 (map-get? tally tally-key))))
+            (total-weight (try! (get-total-weight reward-cycle))))
+        ;; Check that the key has not yet been set for this reward cycle
+        (asserts! (is-none (map-get? aggregate-public-keys reward-cycle)) (err ERR_OUT_OF_VOTING_WINDOW))
+        ;; Check that the aggregate public key is the correct length
+        (asserts! (is-eq (len key) u33) (err ERR_ILL_FORMED_AGGREGATE_PUBLIC_KEY))
+        ;; Check that aggregate public key has not been used in a previous reward cycle
+        (asserts! (is-valid-aggregate-public-key key reward-cycle) (err ERR_DUPLICATE_AGGREGATE_PUBLIC_KEY))
+        ;; Check that signer hasn't voted in this reward-cycle & round
+        (asserts! (map-insert votes {reward-cycle: reward-cycle, round: round, signer: tx-sender} {aggregate-public-key: key, signer-weight: signer-weight}) (err ERR_DUPLICATE_VOTE))
+        ;; Check that the round is incremented by at most 1
+        (try! (update-last-round reward-cycle round))
```

Comment:
I'm not sure this behavior is what we want.  Suppose two sets of signers are participating in two distinct rounds A and B (A < B w.l.o.g.) in the same reward cycle.  Suppose we get the following votes in rounds A and B:

* A sequence of votes for key K_A that put it at 69% agreement.  The current round is A.
* A vote for key K_B that puts it at 1% agreement.  Now, the current round is B.
* A vote for key K_A that puts it over 70% agreement.  The vote will fail because A < B.

In other words, the code here allows a single malicious signer to deny service to voting by posting incrementally larger rounds over the prepare phase.

The property we want from this function is that the aggregate public key K is chosen as the aggregate public key for the given `reward-cycle` if it is the first key in which signers reach 70% agreement with a fixed round number.  Then, the scenario above doesn't happen; key K_A gets selected when the last vote in that list arrives.  It doesn't matter what the round number is, or what value it has relative to other ongoing rounds.  Also, round numbers can be reused across reward cycles.  All the round number is doing is identifying a DKG view that is shared by a bloc of signers.

Reply:
I think this is obscured by the `match/assert/and` logic in `update-last-round`.

If `round` is <= `last_round + 1`, then that function returns `(ok changed?)` where `changed?` is whether or not `last_round` got updated. The `try!` here only returns if there was an error, which only happens if the round is `> last_round + 1`. So, I think the desired behavior is here -- votes can accumulate for any round. Each signer can only vote once per round. The first key that crosses the threshold, regardless of round, is set as the key.

Reply:
Added a test for this in 712c77e05